src/ucs/datastruct/piecewise_func.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2024. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_PIECEWISE_FUNC_H_
+#define UCS_PIECEWISE_FUNC_H_
+
+#include <ucs/datastruct/linear_func.h>
+#include <ucs/datastruct/list.h>
+#include <stddef.h>
+
+
+/*
+ * A piecewise func segment which represents linear function on the range.
+ * Start of the segment equals end of the previous segment + 1. The first
+ * segment starts at 0.
+ */
+typedef struct {
+    ucs_linear_func_t func;
+    size_t            end;
+    ucs_list_link_t   list;
+} ucs_piecewise_segment_t;
+
+
+/**
+ * A piecewise function consisting of one or more segments.
+ */
+typedef struct {
+    ucs_list_link_t head;
+} ucs_piecewise_func_t;
+
+
+/**
+ * Initialize a piecewise function which represents f(x) = 0 on the
+ * [0, SIZE_MAX] range.
+ *
+ * @param [in] func    Piecewise function to initialize.
+ *
+ * @return UCS_OK in case of success, error otherwise.
+ */
+ucs_status_t ucs_piecewise_func_init(ucs_piecewise_func_t *func);
+
+
+/**
+ * Free a piecewise function allocated memory.
+ *
+ * @param [in] func    Piecewise function to free.
+ */
+void ucs_piecewise_func_cleanup(ucs_piecewise_func_t *func);
+
+
+/**
+ * Calculate the piecewise function value for a specific point.
+ *
+ * @param [in] func    Piecewise function to apply.
+ * @param [in] x       Point to apply the function at.
+ *
+ * @return The value of the piecewise function at the given point.
+ */
+double ucs_piecewise_func_apply(const ucs_piecewise_func_t *func, size_t x);
+
+
+/**
+ * Add the trend on the provided segment to the piecewise function.
+ *
+ * @param [inout]  func    Piecewise function to update.
+ * @param [in]     start   Start of the range(inclusive).
+ * @param [in]     end     End of the range(inclusive).
+ * @param [in]     trend   Trend to add.
+ *
+ * @return UCS_OK in case of success, error otherwise.
+ */
+ucs_status_t ucs_piecewise_func_add_segment(ucs_piecewise_func_t *func,
```

Comment:
the name seems a bit confusing, as you may just update existing segment/s (not necessarily to add a new one)

Reply:
Agree, WDYTA `ucs_piecewise_func_sum_up_segment`? 

I also thought about using "enlarge" verb, but seems it doesn't perfectly fit the purpose of the function.

Reply:
Or maybe just change the `_segment` part, and change it to `ucs_piecewise_func_add_trend` or something like that?

Reply:
`add` something which is not a segment sounds good to me